pkg/controller/controller_multiplex.go

```diff
@@ -0,0 +1,175 @@
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+
+	"github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/v1beta1"
+	"github.com/kubernetes-incubator/service-catalog/pkg/pretty"
+	corev1 "k8s.io/api/core/v1"
+
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// Conditions
+
+// ConditionType can be translated to a ServiceInstanceCondition or
+// ServiceBindingCondition value.
+type ConditionType string
+
+const (
+	// ConditionReady represents that a given InstanceCondition is in
+	// ready state.
+	ConditionReady ConditionType = "Ready"
+
+	// ConditionFailed represents information about a final failure
+	// that should not be retried.
+	ConditionFailed ConditionType = "Failed"
+)
+
+// setCondition sets a single condition on a type's status. This delegates to
+// the correct type implementation.
+//
+// Note: objects coming from informers should never be mutated; always pass a
+// deep copy as the instance parameter.
+func setCondition(obj interface{},
+	condition ConditionType,
+	status v1beta1.ConditionStatus,
+	reason,
+	message string) {
+
+	switch toUpdate := obj.(type) {
+	case *v1beta1.ServiceInstance:
+		var conditionType v1beta1.ServiceInstanceConditionType
+		switch condition {
+		case ConditionReady:
+			conditionType = v1beta1.ServiceInstanceConditionReady
+		case ConditionFailed:
+			conditionType = v1beta1.ServiceInstanceConditionReady
+		}
+		setServiceInstanceCondition(toUpdate, conditionType, status, reason, message)
+		return
+	case *v1beta1.ServiceBinding:
+		var conditionType v1beta1.ServiceBindingConditionType
+		switch condition {
+		case ConditionReady:
+			conditionType = v1beta1.ServiceBindingConditionReady
+		case ConditionFailed:
+			conditionType = v1beta1.ServiceBindingConditionFailed
+		}
+		setServiceBindingCondition(toUpdate, conditionType, status, reason, message)
+		return
+	default:
+		glog.Errorf("attempting to set condition on object of unknown type: %+v", obj)
+	}
+}
+
+// Status
+
+// updateStatus inspects then updates status based on type
+func (c *controller) updateStatus(obj interface{}, pcb *pretty.ContextBuilder) (interface{}, error) {
+	switch toUpdate := obj.(type) {
+	case *v1beta1.ServiceInstance:
+		return c.updateServiceInstanceStatus(toUpdate)
+	case *v1beta1.ServiceBinding:
+		return c.updateServiceBindingStatus(toUpdate)
+	default:
+		s := fmt.Sprintf(
+			"attempting to set status on object of unknown type: %+v",
+			obj,
+		)
+		glog.Errorf(pcb.Message(s))
+		return obj, fmt.Errorf(s)
+	}
+}
+
+// prepareInProgressProperties generates the required properties for setting
+// the in-progress status of a Type
+func (c *controller) prepareInProgressProperties(object runtime.Object, toUpdate interface{}, namespace string, specParameters *runtime.RawExtension, specParametersFrom []v1beta1.ParametersFromSource, pcb *pretty.ContextBuilder) (map[string]interface{}, string, *runtime.RawExtension, error) {
```

Comment:
I don't like the name of this function. The function does not actually prepare the in-progress properties; it prepares the parts of the in-progress properties related to parameters.